core/sdk/src/extn/client/extn_client.rs

```diff
@@ -233,119 +259,145 @@ impl ExtnClient {
-                        trace!("fwding event result: {:?}", send_response);
-                    } else {
-                        error!("unable to get sender for target: {:?}", message.target_id);
-                        self.handle_no_processor_error(message);
-                    }
-                } else {
-                    if !is_main {
-                        if let Some(context) = RippleContext::is_ripple_context(&message.payload) {
-                            trace!(
-                                "Received ripple context in {} message: {:?}",
-                                self.sender.get_cap(),
-                                message
-                            );
-                            {
-                                let mut ripple_context = self.ripple_context.write().unwrap();
-                                ripple_context.deep_copy(context);
+            loop {
+                tokio::select! {
+                    Some(value) = &mut read_pin => {
+                        match value {
+                            Ok(msg) => {
+                                if let Message::Text(message) = msg.clone() {
+                                    if let Ok(extn_message) = ExtnMessage::try_from(message) {
+                                        if let Some(ts) = extn_message.ts {
+                                            let latency = Utc::now().timestamp_millis() - ts;
+                                            if latency > 1000 {
+                                                error!("IEC Latency {:?}", msg);
+                                            }
+                                        }
+                                        self.handle_message(extn_message);
+                                    } else {
+                                        error!("Failed to parse message: {:?}", msg);
+                                    }
+                                }
                             }
-                            if !self.has_event_listener(&message.target.as_clear_string()) {
-                                continue;
+                            Err(e) => {
+                                error!("Service Websocket error on read {:?}", e);
+                                break;
                             }
                         }
+                    },
+                    Some(request) = tr.recv() => {
+                        trace!("IEC send: {:?}", request.jsonrpc_msg);
+                        let _feed = ws_tx.feed(Message::Text(request.jsonrpc_msg)).await;
+                        let _flush = ws_tx.flush().await;
                     }
-                    Self::handle_vec_stream(message, self.event_processors.clone());
+                }
+            }
+        }
+        debug!("Initialize Ended Abruptly");
+    }
+
+    pub fn handle_message(&self, message: ExtnMessage) -> ControlFlow<()> {
```

Comment:
The function could be more structured and readable if the message processing were organized as

 ```
   pub fn handle_message(&self, message: ExtnMessage) -> ControlFlow<()> {
        trace!("IEC recv: {:?}", message);

        if message.payload.is_response() {
            return self.handle_response_message(message);
        } else if message.payload.is_event() {
            return self.handle_event_message(message);
        } else {
            return self.handle_request_message(message);
        }
    }
```

Reply:
Good enhancement for future